### PR TITLE
implementation of basic load based segmenet assignment strategy

### PR DESCRIPTION
**Pull request by sajavadi**
Hi,

As we know Segment Assignment Strategy (SAS) can impact end to end response times considerably. Therefore, in order to implement a better SAS we should take computational and storage cost imposed by a segment  to a Pinot server into account. This pull request is the implementation of a generic load based SAS that can use segment_size and segment_count to balance the load among servers. Later on we can add more sophisticated load metrics which lead to more balanced load among servers. 

Best,
Ahmad

### CODE REVIEW

**Comment by raviarin** [comment]
I dont see any test cases in the PR?

________________________________

File: pinot-common/src/main/java/com/linkedin/pinot/common/restlet/resources/ServerSegmentInfo.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class ServerSegmentInfo {
+  private String _serverName;
+  private long _segmentCount;
+  private long _segmentSizeInBytes;
+
+  public ServerSegmentInfo(String serverName) {
+    this._serverName = serverName;
```
Comment (jfim) [inline pinot-common/src/main/java/com/linkedin/pinot/common/restlet/resources/ServerSegmentInfo.java]: Remove `this.`

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/sharding/BalancedLoadAssignmentStrategy.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+It then asks all tagged instances to return the load metric using the ServerLoadMetric object.
+Finally numReplicas of instances that have the least load are selected.
+ */
+public class BalancedLoadAssignmentStrategy implements SegmentAssignmentStrategy {
+  private static final Logger LOGGER = LoggerFactory.getLogger(BalancedLoadAssignmentStrategy.class);
+  private final double MAX_ACCEPTABLE_ERROR_RATE = 0.5;
+  ServerLoadMetric _serverLoadMetric;
+
+  public BalancedLoadAssignmentStrategy(ServerLoadMetric serverLoadMetric) {
+    this._serverLoadMetric = serverLoadMetric;
```
Comment (jfim) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/sharding/BalancedLoadAssignmentStrategy.java]: Remove `.this`

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/util/ServerPerfMetricsReader.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+public class ServerPerfMetricsReader {
+  private static Logger LOGGER = LoggerFactory.getLogger(TableSizeReader.class);
+  private Executor _executor;
+  private HttpConnectionManager _connectionManager;
+  private PinotHelixResourceManager _helixResourceManager;
+
+  public ServerPerfMetricsReader(Executor executor, HttpConnectionManager connectionManager,
+      PinotHelixResourceManager helixResourceManager) {
+    this._executor = executor;
```
Comment (jfim) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/util/ServerPerfMetricsReader.java]: Remove `this.`

**Comment by npawar** [comment]
Closed this PR due to 6 months inactivity. Reopen if needed.

**Pull request closed without merging**